rclcpp/test/benchmark/benchmark_executor.cpp

```diff
@@ -0,0 +1,309 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+
+#include "performance_test_fixture/performance_test_fixture.hpp"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/scope_exit.hpp"
+#include "test_msgs/msg/empty.hpp"
+
+using namespace std::chrono_literals;
+using performance_test_fixture::PerformanceTest;
+
+class PerformanceTestExecutor : public PerformanceTest
+{
+public:
+  void SetUp(benchmark::State & st)
+  {
+    rclcpp::init(0, nullptr);
+    node = std::make_shared<rclcpp::Node>("my_node");
+
+    publisher = node->create_publisher<test_msgs::msg::Empty>("empty_msgs", rclcpp::QoS(10));
+    auto callback = [this](test_msgs::msg::Empty::SharedPtr) {this->callback_count++;};
+    subscription =
+      node->create_subscription<test_msgs::msg::Empty>(
+      "empty_msgs", rclcpp::QoS(10), std::move(callback));
+
+    PerformanceTest::SetUp(st);
+  }
+  void TearDown(benchmark::State & st)
+  {
+    PerformanceTest::TearDown(st);
+    publisher.reset();
+    subscription.reset();
+    node.reset();
+    rclcpp::shutdown();
+  }
+
+  rclcpp::Node::SharedPtr node;
+  rclcpp::Publisher<test_msgs::msg::Empty>::SharedPtr publisher;
+  rclcpp::Subscription<test_msgs::msg::Empty>::SharedPtr subscription;
+  int callback_count;
```

Comment:
This needs to be initialized (either on construction or in setup).

Reply:
it's initialized here:

https://github.com/ros2/rclcpp/blob/6873a47ff379c52246cd5c6295994376d090a430/rclcpp/test/benchmark/benchmark_executor.cpp#L33-L39

Reply:
this->callback_count is not set to 0 as far as I can tell. Or am I missing something?

Reply:
I was planning to check this number when the test is finished to raise an error if there is no message published. (like right now)

Reply:
callback_count still needs to be initialized either during construction or setup of the unit test. Otherwise it's value is undefined and could potentially not be 0.